Readme.md
# Introduction

ImageProcessor is a simple image processor that generates frammed images from a given source directory.

# Using

To use the application, just give the source and target directories.

If you wish to protect your images with a copyright message, you can also give one.

Finally set how much to keep from the original size from the source images.

The generated images will be framed with the copyright message on the lower right corner.

# Building

This application was developed in Java. A Netbeans project definition file is also available.

Changes.txt
======== V 0.0.1 Release was released ===========================

2006       (first release) : Made the files available on my own web site. (Paulo Pinto)
2012-11-25 (second release): Pushed the files into GitHub. (Paulo Pinto)


src/org/progtools/ProgressDlg.java
/*
 * ProgressDlg.java
 * Shows the progress of the images being processed.
 * Copyright (C) 2006  Paulo Pinto
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.io.File;

/**
 * To be called to make the user aware that something is happening.
 * @author  Paulo Pinto
 */
public class ProgressDlg extends javax.swing.JDialog implements ImageProcessorListener {
    
    /**
     * Creates new form ProgressDlg
     */
    public ProgressDlg(java.awt.Frame parent, boolean modal) {
        super(parent, modal);
        initComponents();
        
         // Center the application
        setLocationRelativeTo(null);
    }
    
    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc=" Generated Code ">//GEN-BEGIN:initComponents
    private void initComponents() {
        msgLbl = new javax.swing.JLabel();

        setDefaultCloseOperation(javax.swing.WindowConstants.DISPOSE_ON_CLOSE);
        setTitle("Progess");
        setName("progressDlg");
        setResizable(false);
        getAccessibleContext().setAccessibleParent(null);
        msgLbl.setText("Processing");

        org.jdesktop.layout.GroupLayout layout = new org.jdesktop.layout.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(layout.createSequentialGroup()
                .add(165, 165, 165)
                .add(msgLbl)
                .addContainerGap(184, Short.MAX_VALUE))
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(org.jdesktop.layout.GroupLayout.TRAILING, layout.createSequentialGroup()
                .addContainerGap(org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                .add(msgLbl, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 25, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                .addContainerGap())
        );
        pack();
    }// </editor-fold>//GEN-END:initComponents
    
    /**
     * Shows the current image being processed.
     * @param pathname The name of the image that is being processed.
     */
    public void processedImage(File pathname) {
        msgLbl.setText(pathname.getName());
    }
    
    /**
     * Just for testing purposes.
     * @param args the command line arguments
     */
    public static void main(String args[]) {
        java.awt.EventQueue.invokeLater(new Runnable() {
            public void run() {
                new ProgressDlg(new javax.swing.JFrame(), true).setVisible(true);
            }
        });
    }
    
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JLabel msgLbl;
    // End of variables declaration//GEN-END:variables
    
}


src/org/progtools/ImageProcessor.java
/*
 * ImageProcessor.java
 * An image processor to generate frammed images with copyright messages
 * Copyright (C) 2006  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.awt.Color;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import javax.imageio.ImageIO;
import javax.swing.SwingUtilities;

/**
 * A class that processes images to be placed in a
 * Website photo gallery, with frames and copyright
 * messages.
 */
public class ImageProcessor {
    private static final int FRAME_WIDTH = 20;
    
    private String destinationDir;
    private String copyrightMsg;
    
    private List<ImageProcessorListener> listeners;
    
    /**
     * Creates an image processor.
     * @param destinationDir A pathname to the directory where to place the generated images
     * @param copyrightMsg The copyright message to place in the images
     */
    public ImageProcessor(String destinationDir, String copyrightMsg) {
        this.destinationDir = destinationDir;
        this.copyrightMsg = copyrightMsg;
        this.listeners = new ArrayList<ImageProcessorListener>(10);
    }
    
    /**
     * Converts the given image to the dimensions specified in the constructor.
     * An image frame is added, as well as the copyright message.
     * @param sourceImage The image to convert.
     * @param size A value between 0 and 100 to say how much we need to reduce the image
     */
    public void convertImage(File sourceImage, int size) {
        assert size >= 0 && size <= 100;
        try {
            // Load the images
            BufferedImage originalImg = ImageIO.read(sourceImage);
            
            
            // Calculate the size for the new image
            int width = originalImg.getWidth();
            int height = originalImg.getHeight();
            if (width > height) {
                float ratio = (float) height / width;
                width *= ((float)size / 100);
                height = (int) (ratio * width);
            } else {
                float ratio = (float) width / height;
                height *= ((float)size / 100);
                width = (int) (ratio * height);
            }
            BufferedImage convertedImg = new BufferedImage(width + FRAME_WIDTH * 2,
                    height + FRAME_WIDTH * 2, originalImg
                    .getType());
            
            // Convert into a new one with frames
            Graphics2D g = (Graphics2D) convertedImg.getGraphics();
            g.setColor(Color.WHITE);
            g.fillRect(0, 0, convertedImg.getWidth(), convertedImg.getHeight());
            
            g.setColor(Color.BLACK);
            g.drawRect(0, 0, convertedImg.getWidth(), convertedImg.getHeight());
            
            g.drawImage(originalImg, FRAME_WIDTH, FRAME_WIDTH, convertedImg
                    .getWidth() - (FRAME_WIDTH * 2), convertedImg.getHeight() - (FRAME_WIDTH * 2), null);
            
            
            // Add the copyright message
            // In the future I might change this font for a font effect like
            // embossed text.
            Font fnt = new Font("Arial", Font.BOLD, 16);
            g.setFont(fnt);
            FontMetrics metrics = g.getFontMetrics(fnt);
            int hgt = metrics.getHeight();
            int adv = metrics.stringWidth(copyrightMsg) + 5; // The 5 is a prefix padding for the text
            
            //g.setXORMode(Color.WHITE);
            g.setColor(Color.WHITE);
            g.drawString(copyrightMsg, convertedImg.getWidth() - FRAME_WIDTH - adv, convertedImg.getHeight() - FRAME_WIDTH - hgt);
            
            // Save the converted image
            String destPath = destinationDir + File.separator + sourceImage.getName();
            ImageIO.write(convertedImg, "jpg", new File(destPath));
            notifyListeners(sourceImage);
        } catch (IOException e) {
            System.out.println("An error occurred while converting the image: " +  e.getMessage());
        }
    }
    
    /**
     * Processes all the image files from the given directory.
     * The images will be generated with the same name in the destination
     * directory given in the constructor.
     * @param pathname A pathname for a directory
     * @param size A value between 0 and 100 to say how much we need to reduce the images
     */
    public void processFiles(String pathname, int size) {
        File imagesDir = new File(pathname);
        for (File image: imagesDir.listFiles()) {
            if (image.getName().toLowerCase().endsWith(".jpg")) {
                convertImage(image, size);
            }
        }
    }
    
    /**
     * Notifies all listeners of the processed image.
     * It makes use of SwingUtilities so that Swing code can
     * also be notified of the progress.
     * @param imageFile The image that was just processed.
     */
    private void notifyListeners(final File imageFile) {
        // Make sure all listeners are updated in a thread safe way
        try {
            Runnable notifier = new Runnable() {
                public void run() {
                    for (ImageProcessorListener listener : listeners)
                        listener.processedImage(imageFile);
                }
            };
            SwingUtilities.invokeAndWait(notifier);
        } catch (Exception e) {
            System.out.println("Error while invonking listeners: " + e.getMessage());
        }
    }
    
    /**
     * Adds a new listener client to this class
     * @param listener The listener to add.
     */
    public void addImageProcessorListener(ImageProcessorListener listener) {
        listeners.add(listener);
    }
    
    /**
     * Removes a listener client to this class
     * @param listener The listener to add.
     */
    public void removeImageProcessorListener(ImageProcessorListener listener) {
        listeners.remove(listener);
    }
    
}



src/org/progtools/ImageProcessorListener.java
/*
 * ImageProcessorListener.java
 * Listener interface for the image processor
 * Copyright (C) 2006  Paulo Pinto
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.io.File;

/**
 * Listener interface for classes interested in image processing
 * events.
 * @author Paulo
 */
public interface ImageProcessorListener {
    
    /**
     * Invoked when an imaged gets processed by the image processor
     * @param pathname Pathname to the processed image.
     */
    void processedImage(File pathname);
}


src/org/progtools/GuiApp.java
/*
 * GuiApp.java
 * Main Application Window
 * Copyright (C) 2006  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.awt.HeadlessException;
import javax.swing.JFileChooser;
import javax.swing.JTextField;
import javax.swing.UIManager;

/**
 * Application main window.
 * @author  Paulo Pinto
 */
public class GuiApp extends javax.swing.JFrame {
    
    /**
     * Creates the main application frame.
     */
    public GuiApp() {
        initComponents();
        
        // Center the application window
        setLocationRelativeTo(null);
    }
    
    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc=" Generated Code ">//GEN-BEGIN:initComponents
    private void initComponents() {
        jLabel1 = new javax.swing.JLabel();
        jLabel2 = new javax.swing.JLabel();
        jLabel3 = new javax.swing.JLabel();
        jLabel4 = new javax.swing.JLabel();
        srcDirTxt = new javax.swing.JTextField();
        destDirTxt = new javax.swing.JTextField();
        msgTxt = new javax.swing.JTextField();
        imgSizeSlider = new javax.swing.JSlider();
        srcBrowseBtn = new javax.swing.JButton();
        destBrowseBtn = new javax.swing.JButton();
        quitBtn = new javax.swing.JButton();
        processBtn = new javax.swing.JButton();
        jSeparator1 = new javax.swing.JSeparator();

        setDefaultCloseOperation(javax.swing.WindowConstants.EXIT_ON_CLOSE);
        setTitle("Select the Images to Process");
        jLabel1.setText("Source Directory:");

        jLabel2.setText("Target Directory:");

        jLabel3.setText("Copyright Message:");

        jLabel4.setText("Image Size:");

        imgSizeSlider.setMajorTickSpacing(10);
        imgSizeSlider.setPaintLabels(true);
        imgSizeSlider.setPaintTicks(true);

        srcBrowseBtn.setText("Browse...");
        srcBrowseBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                srcBrowseBtnActionPerformed(evt);
            }
        });

        destBrowseBtn.setText("Browse...");
        destBrowseBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                destBrowseBtnActionPerformed(evt);
            }
        });

        quitBtn.setText("Quit");
        quitBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                quitBtnActionPerformed(evt);
            }
        });

        processBtn.setText("Process");
        processBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                processBtnActionPerformed(evt);
            }
        });

        org.jdesktop.layout.GroupLayout layout = new org.jdesktop.layout.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(layout.createSequentialGroup()
                .add(36, 36, 36)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(jLabel1)
                    .add(jLabel2)
                    .add(jLabel3)
                    .add(jLabel4))
                .add(10, 10, 10)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING, false)
                    .add(imgSizeSlider, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .add(msgTxt)
                    .add(destDirTxt)
                    .add(srcDirTxt, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, 240, Short.MAX_VALUE))
                .add(28, 28, 28)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.TRAILING)
                    .add(srcBrowseBtn)
                    .add(destBrowseBtn))
                .addContainerGap())
            .add(org.jdesktop.layout.GroupLayout.TRAILING, layout.createSequentialGroup()
                .addContainerGap()
                .add(jSeparator1, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 473, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                .addContainerGap())
            .add(org.jdesktop.layout.GroupLayout.TRAILING, layout.createSequentialGroup()
                .addContainerGap(361, Short.MAX_VALUE)
                .add(processBtn)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(quitBtn)
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(layout.createSequentialGroup()
                .add(29, 29, 29)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(jLabel1)
                    .add(srcBrowseBtn)
                    .add(srcDirTxt, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE))
                .add(24, 24, 24)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(jLabel2)
                    .add(destDirTxt, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                    .add(destBrowseBtn))
                .add(31, 31, 31)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(jLabel3)
                    .add(msgTxt, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE))
                .add(20, 20, 20)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(jLabel4)
                    .add(imgSizeSlider, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(jSeparator1, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 34, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(quitBtn)
                    .add(processBtn))
                .addContainerGap(org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );
        pack();
    }// </editor-fold>//GEN-END:initComponents
    
    /**
     * Starts the processing of the images.
     * @param evt The swing event for the pressed button
     */
    private void processBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_processBtnActionPerformed
        final ProgressDlg dialog = new ProgressDlg(this, false);
        dialog.setVisible(true);
        
        final ImageProcessor processor = new ImageProcessor(destDirTxt.getText(), msgTxt.getText());
        processor.addImageProcessorListener(dialog);
        
        // Run this in a background thread so that the UI is
        // free to process events
        Thread processorThread = new Thread() {
            public void run() {
                processor.processFiles(srcDirTxt.getText(), imgSizeSlider.getValue());
                processor.removeImageProcessorListener(dialog);
                dialog.dispose();
            }
        };
        processorThread.start();
    }//GEN-LAST:event_processBtnActionPerformed
    
    /**
     * Allows the user to select the destination directory for the images.
     * @param evt The swing event for the pressed button
     */
    private void destBrowseBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_destBrowseBtnActionPerformed
        browseDirectory(destDirTxt);
    }//GEN-LAST:event_destBrowseBtnActionPerformed
    
    /**
     * Allows the user to select the source directory for the images.
     * @param evt The swing event for the pressed button
     */
    private void srcBrowseBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_srcBrowseBtnActionPerformed
        browseDirectory(srcDirTxt);
    }//GEN-LAST:event_srcBrowseBtnActionPerformed
    
    /**
     * Shows the file open dialog box for the user to choose a directory.
     * If one is selected, the pathname is assigned to the given field.
     * @param field Where to assign the pathname of the chosen directory
     */
    private void browseDirectory(JTextField field) throws HeadlessException {
        JFileChooser chooser = new JFileChooser();
        chooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
        int returnVal = chooser.showOpenDialog(this);
        if(returnVal == JFileChooser.APPROVE_OPTION) {
            field.setText(chooser.getSelectedFile().getPath());
        }
    }
    
    /**
     * Releases the application, thus closing it.
     * @param evt The swing event for the pressed button
     */
    private void quitBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_quitBtnActionPerformed
        dispose();
    }//GEN-LAST:event_quitBtnActionPerformed
    
    /**
     * Standard main application entry point
     * @param args the command line arguments
     */
    public static void main(String args[]) {
        try {
            // Set System L&F
            UIManager.setLookAndFeel(
                    UIManager.getSystemLookAndFeelClassName());
        } catch (Exception e) {
            System.out.println("Error while trying to set the system look and feel! : " + e.getMessage());
        }
        java.awt.EventQueue.invokeLater(new Runnable() {
            public void run() {
                new GuiApp().setVisible(true);
            }
        });
    }
    
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JButton destBrowseBtn;
    private javax.swing.JTextField destDirTxt;
    private javax.swing.JSlider imgSizeSlider;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JLabel jLabel2;
    private javax.swing.JLabel jLabel3;
    private javax.swing.JLabel jLabel4;
    private javax.swing.JSeparator jSeparator1;
    private javax.swing.JTextField msgTxt;
    private javax.swing.JButton processBtn;
    private javax.swing.JButton quitBtn;
    private javax.swing.JButton srcBrowseBtn;
    private javax.swing.JTextField srcDirTxt;
    // End of variables declaration//GEN-END:variables
}
